### PR TITLE
Backport of LHEGenericMassFilter to 106X

### PR DESCRIPTION
**Pull request by JanFSchulte**
This PR backports the LHE filter designed to cut on the invariant mass of a number of particles introduced in #35938. Backport to 10_6_X is necessary since some high dilepton mass ttbar and WW samples need to be produced in the UL campaigns.

### CODE REVIEW

**Comment by agrohsje** [comment]
Hi @qliphy @perrotta , I removed myself from the l2-generators team but I still get mails about PR's. Does it take a while to update or should I remove my name somewhere else?

**Comment by agrohsje** [comment]
please test

**Comment by perrotta** [comment]
@agrohsje I still see your nick listed for "generators" in https://github.com/cms-sw/cms-bot/blob/master/categories.py

The following bot PR removes it: 
https://github.com/cms-sw/cms-bot/pull/1692
If it is actually what you wan't, please confirm and we'll merge it.

**Comment by JanFSchulte** [comment]
I am not sure what the issue is here. In the tests I see only workflow 4.76 failing, which I can not check locally since the input files are not found at T2_CH_CERN. The code introduced here is not used in any of the tests, so I can't see how a failure would be related.

**Comment by perrotta** [comment]
> I am not sure what the issue is here. In the tests I see only workflow 4.76 failing, which I can not check locally since the input files are not found at T2_CH_CERN. The code introduced here is not used in any of the tests, so I can't see how a failure would be related.

@JanFSchulte the crash in wf 4.76 is unrelated (see https://github.com/cms-sw/cmssw/issues/36771). Please disregard that workflow for now

**Comment by SiewYan** [comment]
@perrotta , should i re-trigger the test and avoid 4.76 to complete the test?

**Comment by JanFSchulte** [comment]
Could someone comment on the status of this PR? We would like to have if merged so we know we can move ahead with preparing the sample requests.

File: GeneratorInterface/GenFilters/interface/LHEGenericMassFilter.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#include "SimDataFormats/GeneratorProducts/interface/LHEEventProduct.h"
+
+//
+// class declaration
+//
+
+class LHEGenericMassFilter : public edm::global::EDFilter<> {
+public:
+  explicit LHEGenericMassFilter(const edm::ParameterSet&);
+  ~LHEGenericMassFilter() override;
```
Comment (perrotta) [inline GeneratorInterface/GenFilters/interface/LHEGenericMassFilter.h:35]:
```suggestion
  ~LHEGenericMassFilter() override = default;
```

File: GeneratorInterface/GenFilters/src/LHEGenericMassFilter.cc [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    : src_(consumes<LHEEventProduct>(iConfig.getParameter<edm::InputTag>("src"))),
+      numRequired_(iConfig.getParameter<int>("NumRequired")),
+      particleID_(iConfig.getParameter<std::vector<int> >("ParticleID")),
+      minMass_(iConfig.getParameter<double>("MinMass")),
+      maxMass_(iConfig.getParameter<double>("MaxMass")) {}
+
+LHEGenericMassFilter::~LHEGenericMassFilter() {
+  // do anything here that needs to be done at destruction time
+  // (e.g. close files, deallocate resources etc.)
+}
```
Comment (perrotta) [inline GeneratorInterface/GenFilters/src/LHEGenericMassFilter.cc:16]:
```suggestion
```

File: GeneratorInterface/GenFilters/interface/LHEGenericMassFilter.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+private:
+  bool filter(edm::StreamID, edm::Event&, edm::EventSetup const&) const override;
+  void endJob() override;
```
Comment (perrotta) [inline GeneratorInterface/GenFilters/interface/LHEGenericMassFilter.h:39]: Probably not needed

File: GeneratorInterface/GenFilters/src/LHEGenericMassFilter.cc [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        break;  // only match a given particle once!
+      }
+    }  // loop over targets
+
+  }  // loop over particles
+
+  // event accept/reject logic
+  if (nFound == numRequired_) {
+    double Mass = std::sqrt(E * E - (Px * Px + Py * Py + Pz * Pz));
+    if (Mass > minMass_ && Mass < maxMass_) {
```
Comment (perrotta) [inline GeneratorInterface/GenFilters/src/LHEGenericMassFilter.cc:51]:
```suggestion
    double sqrdMass = (E * E - (Px * Px + Py * Py + Pz * Pz);
    if (sqrdMass > minMass_*minMass_ && Mass < maxMass_*maxMass_) {
```

**Comment by perrotta** [comment]
As in the backported PR, I would move it into the \plugin area, also combining the header and implementation file into a single .cc
By the way, why didn't you backport it as it was?

File: GeneratorInterface/GenFilters/interface/LHEGenericMassFilter.h (hunk 1 above)
Comment (JanFSchulte) [reply GeneratorInterface/GenFilters/interface/LHEGenericMassFilter.h]: Done

File: GeneratorInterface/GenFilters/src/LHEGenericMassFilter.cc (hunk 2 above)
Comment (JanFSchulte) [reply GeneratorInterface/GenFilters/src/LHEGenericMassFilter.cc]: Done

File: GeneratorInterface/GenFilters/interface/LHEGenericMassFilter.h (hunk 3 above)
Comment (JanFSchulte) [reply GeneratorInterface/GenFilters/interface/LHEGenericMassFilter.h]: Removed.

File: GeneratorInterface/GenFilters/src/LHEGenericMassFilter.cc (hunk 4 above)
Comment (JanFSchulte) [reply GeneratorInterface/GenFilters/src/LHEGenericMassFilter.cc]: Done.

**Comment by JanFSchulte** [comment]
I have merged the files and moved them to plugins. I had split them up to conform to the structure of the other modules in this package, and because I had issues getting it to be included in the compilation when it was in the plugins folder. I realized now that was because there was no BuildFile.xml there, so now everything works.

**Comment by SiewYan** [comment]
please test

**Comment by SiewYan** [comment]
+1

its look good to me.

**Comment by qliphy** [comment]
+1

**Pull request merged**